jmx/jmx.go

```diff
@@ -39,97 +39,34 @@ const (
-	var cliCommand []string
-
+func getCommand(hostname, port, username, password string) (c []string) {
 	if os.Getenv("NR_JMX_TOOL") != "" {
-		cliCommand = strings.Split(os.Getenv("NR_JMX_TOOL"), " ")
+		c = strings.Split(os.Getenv("NR_JMX_TOOL"), " ")
 	} else {
-		cliCommand = []string{jmxCommand}
+		c = []string{jmxCommand}
 	}
 
-	cliCommand = append(cliCommand, "--hostname", hostname, "--port", port)
+	c = append(c, "--hostname", hostname, "--port", port)
 	if username != "" && password != "" {
-		cliCommand = append(cliCommand, "--username", username, "--password", password)
+		c = append(c, "--username", username, "--password", password)
 	}
 
-	return cliCommand
+	return
 }
 
 func getCommandWithSSL(hostname, port, username, password, keyStore, keyStorePassword, trustStore, trustStorePassword string) []string {
-	var cliCommand []string
-
-	if os.Getenv("NR_JMX_TOOL") != "" {
-		cliCommand = strings.Split(os.Getenv("NR_JMX_TOOL"), " ")
-	} else {
-		cliCommand = []string{jmxCommand}
-	}
-
-	cliCommand = append(cliCommand, "--hostname", hostname, "--port", port)
-	if username != "" && password != "" {
-		cliCommand = append(cliCommand, "--username", username, "--password", password)
-	}
+	c := getCommand(hostname, port, username, password)
 
 	if keyStore != "" && keyStorePassword != "" && trustStore != "" && trustStorePassword != "" {
-		cliCommand = append(cliCommand, "--keyStore", keyStore, "--keyStorePassword", keyStorePassword, "--trustStore", trustStore, "--trustStorePassword", trustStorePassword)
+		c = append(c, "--keyStore", keyStore, "--keyStorePassword", keyStorePassword, "--trustStore", trustStore, "--trustStorePassword", trustStorePassword)
 	}
 
-	return cliCommand
+	return c
 }
 
 // Open will start the nrjmx command with the provided connection parameters.
 func Open(hostname, port, username, password string) error {
-	lock.Lock()
-	defer lock.Unlock()
-
-	if cmd != nil {
-		return ErrJmxCmdRunning
-	}
-
-	// Drain error channel to prevent showing past errors
-	if len(cmdErr) > 0 {
-		<-cmdErr
-	}
-
-	done.Add(1)
-
-	var err error
-	var ctx context.Context
-
-	cliCommand := getCommand(hostname, port, username, password)
-
-	ctx, cancel = context.WithCancel(context.Background())
-	cmd = exec.CommandContext(ctx, cliCommand[0], cliCommand[1:]...)
-
-	if cmdOut, err = cmd.StdoutPipe(); err != nil {
-		return err
-	}
-	if cmdIn, err = cmd.StdinPipe(); err != nil {
-		return err
-	}
-
-	if cmdError, err = cmd.StderrPipe(); err != nil {
-		return err
-	}
-
-	if err = cmd.Start(); err != nil {
-		return err
-	}
-
-	go func() {
-		if err = cmd.Wait(); err != nil {
-			stdErr, _ := ioutil.ReadAll(cmdError)
-			cmdErr <- fmt.Errorf("JMX tool exited with error: %s [state: %s] (%s)", err, cmd.ProcessState, string(stdErr))
-		}
-
-		lock.Lock()
-		defer lock.Unlock()
-		cmd = nil
-
-		done.Done()
-	}()
-
-	return nil
+	return OpenWithSSL(hostname, port, username, password, "", "", "", "")
```

Comment:
Methods code are not 100% equal. `Open` invoked `getCommand` while `OpenWithSSL` invokes `getCommandWithSSL`... won't this change break non-ssl connections?

Reply:
If you look at `getCommandWithSSL` it does the same as `getCommand` but only appends ssl args when args are not empty, so calling `Open` is the same as calling always `OpenWithSSL` with empty ssl args